src/logion-chain/LogionClientContext.tsx

```diff
@@ -1,35 +1,128 @@
 import config from '../config/index';
-import { LogionClient, DefaultSignAndSendStrategy } from "@logion/client";
+import { LogionClient, DefaultSignAndSendStrategy, SponsorshipState } from "@logion/client";
 import { enableMetaMask, allMetamaskAccounts, ExtensionSigner } from "@logion/extension";
-import { AnyAccountId } from "@logion/node-api";
-import { Context, createContext, ReactNode, useState, useEffect, useContext } from "react";
+import { UUID, LogionNodeApiClass } from "@logion/node-api";
+import { Context, createContext, ReactNode, useState, useEffect, useContext, useCallback } from "react";
+import { useParams } from "react-router-dom";
 
-export interface LogionClientContextType {
-    client: LogionClient | null
+interface StaticState {
+    client: LogionClient | null;
+    sponsorshipId: UUID | null;
 }
 
-const INITIAL_STATE: LogionClientContextType = {
-    client: null
+export interface DynamicState {
+    sponsorshipState: SponsorshipState | null;
+    errorMessage: string | null;
+    refresh: () => Promise<void>;
 }
-const LogionClientContext: Context<LogionClientContextType> = createContext<LogionClientContextType>(INITIAL_STATE);
+
+export interface LogionClientContextType extends StaticState, DynamicState {
+}
+
+const INITIAL_STATIC_STATE: StaticState = {
+    client: null,
+    sponsorshipId: null,
+}
+
+const INITIAL_DYNAMIC_STATE: DynamicState = {
+    sponsorshipState: null,
+    errorMessage: null,
+    refresh: () => Promise.reject(),
+}
+
+const LogionClientContext: Context<LogionClientContextType> = createContext<LogionClientContextType>({
+    ...INITIAL_STATIC_STATE,
+    ...INITIAL_DYNAMIC_STATE,
+});
 
 export interface Props {
     children: ReactNode;
 }
 
 export default function LogionClientContextProvider(props: Props) {
-    const [ state, setState ] = useState<LogionClientContextType>(INITIAL_STATE);
+    const [ staticState, setStaticState ] = useState<StaticState>(INITIAL_STATIC_STATE);
+    const [ dynamicState, setDynamicState ] = useState<DynamicState>(INITIAL_DYNAMIC_STATE);
+
+    const sponsorshipIdParam = useParams<"sponsorshipId">().sponsorshipId;
+
+    const refresh = useCallback(async (currentSponsorshipState: SponsorshipState) => {
+        console.log("Refreshing...");
+        const newSponsorshipState = await currentSponsorshipState.refresh();
+        console.log("newSponsorshipState: %s", newSponsorshipState.locRequestState?.data().status);
+        setDynamicState({
+            errorMessage: null,
+            sponsorshipState: newSponsorshipState,
+            refresh: () => refresh(newSponsorshipState)
+        })
+    }, [])
+
+    useEffect(() => {
+        (async function () {
+            if (staticState.client === null && dynamicState.errorMessage === null) {
```

Comment:
Move this condition out of the async and add a flag to the condition (telling if the context is initializing or not) that you set before executing the async, otherwise you could get a race condition where the code is executed several times (if the component gets re-rendered while initializing state).

Reply:
Done: https://github.com/logion-network/logion-idloc-gateway/pull/5/commits/7bf7a3e37e98365a90f9163ce472e2a88070dbd9